service/history/handler.go

```diff
@@ -197,11 +196,32 @@ func (h *Handler) isStopped() bool {
 	return atomic.LoadInt32(&h.status) == common.DaemonStatusStopped
 }
 
-// IsWorkflowTaskValid - whether workflow task is still valid
-func (h *Handler) IsWorkflowTaskValid(ctx context.Context, request *historyservice.IsWorkflowTaskValidRequest) (_ *historyservice.IsWorkflowTaskValidResponse, retError error) {
-	defer log.CapturePanic(h.logger, &retError)
+func lifecycleInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (_ interface{}, retErr error) {
+	h, _ := info.Server.(*Handler)
```

Comment:
This is probably safe to do now, but it may not be in the future. However, I think we can do even better than checking the error here. Why don't you make the `lifecycleInterceptor` a method of the handler itself? That way it can check its own `isStopped` state, so we don't have to look into the `info` and validate its type. As long as there are no circular dependencies introduced by making the gRPC options provider depend on the `Handler` type, it should be very easy to do.

Reply:
I like that, let me try that!